services/cargohold/test/integration/API/V3.hs

```diff
@@ -76,45 +76,55 @@ tests s = testGroup "v3"
-        assertEqual "content-type mismatch" (Just applicationText) (getContentType r3)
-        assertEqual "token mismatch" tok (decodeHeader "x-amz-meta-token" r3)
-        assertEqual "user mismatch" uid (decodeHeader "x-amz-meta-user" r3)
-        assertEqual "data mismatch" (Just "Hello World") (responseBody r3)
-
-    -- Delete (forbidden for other users)
-    deleteAsset c uid2 (view V3.assetKey ast) !!! const 403 === statusCode
-    -- Delete (allowed for creator)
-    deleteAsset c uid (view V3.assetKey ast) !!! const 200 === statusCode
-    r4 <- get (c . path loc . zUser uid . header "Asset-Token" (toByteString' tok) . noRedirect) <!!
-        const 404 === statusCode
-    let Just date' = C8.unpack <$> lookup "Date" (responseHeaders r4)
-    let utc' = parseTimeOrError False defaultTimeLocale rfc822DateFormat date' :: UTCTime
-    liftIO $ assertBool "bad date" (utc' >= utc)
+        let      loc = decodeHeader "Location" r1 :: ByteString
+        let Just ast = decodeBody r1 :: Maybe V3.Asset
+        let Just tok = view V3.assetToken ast
+
+        -- Check mandatory Date header
+        let Just date = C8.unpack <$> lookup "Date" (responseHeaders r1)
+        let utc = parseTimeOrError False defaultTimeLocale rfc822DateFormat date :: UTCTime
+        -- Potentially check for the expires header
+        when (isJust $ join (V3.assetRetentionSeconds <$> (sets^.V3.setAssetRetention))) $ do
+            liftIO $ assertBool "invalid expiration" (Just utc < view V3.assetExpires ast)
+
+        -- Lookup with token and download via redirect.
+        r2 <- get (c . path loc . zUser uid . header "Asset-Token" (toByteString' tok) . noRedirect) <!! do
+            const 302 === statusCode
+            const Nothing === responseBody
+        r3 <- flip get' id =<< parseUrlThrow (C8.unpack (getHeader' "Location" r2))
+        liftIO $ do
+            assertEqual "status" status200 (responseStatus r3)
+            assertEqual "content-type mismatch" (Just applicationText) (getContentType r3)
+            assertEqual "token mismatch" tok (decodeHeader "x-amz-meta-token" r3)
+            assertEqual "user mismatch" uid (decodeHeader "x-amz-meta-user" r3)
+            assertEqual "data mismatch" (Just "Hello World") (responseBody r3)
+
+        -- Delete (forbidden for other users)
+        deleteAsset c uid2 (view V3.assetKey ast) !!! const 403 === statusCode
+        -- Delete (allowed for creator)
+        deleteAsset c uid (view V3.assetKey ast) !!! const 200 === statusCode
+        r4 <- get (c . path loc . zUser uid . header "Asset-Token" (toByteString' tok) . noRedirect) <!!
+            const 404 === statusCode
+        let Just date' = C8.unpack <$> lookup "Date" (responseHeaders r4)
+        let utc' = parseTimeOrError False defaultTimeLocale rfc822DateFormat date' :: UTCTime
+        liftIO $ assertBool "bad date" (utc' >= utc)
 
 testSimpleTokens :: TestSignature ()
 testSimpleTokens c = do
```

Comment:
When I read `simpleRoundtrip`, I expected the trivial aeson encode-decode loop and comparison.  Should this test be added for all json-deriving data types nearby, while we're at it?

Reply:
I think that's a good idea, let me deal with that in a separate